pallets/gas/src/lib.rs

```diff
@@ -349,112 +369,54 @@ where
-
-                    if let ValueType::SpecifiedLocal {
-                        value: self_value, ..
-                    } = node.inner
-                    {
-                        if node.unspec_refs == 0 {
-                            // Any node of type `ValueType::SpecifiedLocal` must have a parent
-                            let mut parent_node = Self::get_node(parent)
-                                .ok_or(<Error<T>>::GasTreeInvalidated)?
-                                // Upstream node with a concrete value must exist for any node
-                                .node_with_value::<T>()?;
-
-                            // NOTE: intentional expect. A node_with_value is guaranteed to have inner_value
-                            let parent_val = parent_node
-                                .inner_value_mut()
-                                .expect("Querying parent with value");
-                            *parent_val = parent_val.saturating_add(self_value);
-
-                            GasTree::<T>::mutate(parent_node.id, |value| {
-                                *value = Some(parent_node);
-                            });
-                        }
-                    }
-
-                    if delete_current_node {
-                        GasTree::<T>::remove(key);
-                    } else {
-                        node.consumed = true;
-
-                        if node.unspec_refs == 0 {
-                            if let Some(inner_value) = node.inner_value_mut() {
-                                *inner_value = 0
-                            };
-                        }
-                        // Save current node
-                        GasTree::<T>::mutate(key, |value| {
-                            *value = Some(node);
-                        });
-                    }
-
-                    // now check if the parent node can be consumed as well
-                    if consume_parent_node {
-                        Self::check_consumed(parent)?
-                    } else {
-                        None
-                    }
-                }
-                ValueType::External { id, value } => {
-                    node.consumed = true;
+        let mut node = Self::get_node(key).ok_or(Error::<T>::NodeNotFound)?;
 
-                    if node.refs() == 0 {
-                        // Delete current node
-                        GasTree::<T>::remove(key);
+        ensure!(!node.consumed, Error::<T>::NodeWasConsumed);
 
-                        Some((NegativeImbalance::new(value), id))
-                    } else {
-                        // Save current node
-                        GasTree::<T>::mutate(key, |n| {
-                            *n = Some(node);
-                        });
+        node.consumed = true;
+        node.move_value_upstream::<T>()?;
 
-                        None
-                    }
-                }
+        let outcome = if node.refs() == 0 {
```

Comment:
I'd remove this var to increase readability by directly return Result from if branches

Reply:
done